pkg/controller/nativedaemonset/nativedaemonset_controller.go

```diff
@@ -0,0 +1,267 @@
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	daemonsetutil "github.com/openkruise/rollouts/pkg/controller/nativedaemonset/util"
+	"github.com/openkruise/rollouts/pkg/util"
+	clientutil "github.com/openkruise/rollouts/pkg/util/client"
+)
+
+func init() {
+	flag.IntVar(&concurrentReconciles, "nativedaemonset-workers", concurrentReconciles, "Max concurrent workers for native daemonset controller.")
+}
+
+const (
+	DefaultRetryDuration = 2 * time.Second
+	ControllerName       = "native-daemonset-controller"
+)
+
+var (
+	concurrentReconciles = 3
+)
+
+// Add creates a new Native DaemonSet Controller and adds it to the Manager with default RBAC.
+func Add(mgr manager.Manager) error {
+	r, err := newReconciler(mgr)
+	if err != nil {
+		return err
+	}
+	return add(mgr, r)
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) (reconcile.Reconciler, error) {
+	// Client & Recorder
+	genericClient := clientutil.GetGenericClientWithName(ControllerName)
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(klog.Infof)
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: genericClient.KubeClient.CoreV1().Events("")})
+	recorder := mgr.GetEventRecorderFor(ControllerName)
+
+	return &ReconcileNativeDaemonSet{
+		Client:        mgr.GetClient(),
+		eventRecorder: recorder,
+	}, nil
+}
+
+// ReconcileNativeDaemonSet reconciles a Native DaemonSet object
+type ReconcileNativeDaemonSet struct {
+	// client interface
+	client.Client
+	eventRecorder record.EventRecorder
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(ControllerName, mgr, controller.Options{
+		Reconciler: r, MaxConcurrentReconciles: concurrentReconciles})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to DaemonSet with specific annotations
+	updateHandler := func(e event.UpdateEvent) bool {
+		oldObject := e.ObjectOld.(*appsv1.DaemonSet)
+		newObject := e.ObjectNew.(*appsv1.DaemonSet)
+		if len(oldObject.Annotations) != len(newObject.Annotations) || !reflect.DeepEqual(oldObject.Annotations, newObject.Annotations) {
+			klog.V(3).Infof("Observed updated Annotation for native DaemonSet: %s/%s", newObject.Namespace, newObject.Name)
+			return true
+		}
+		return false
+	}
+
+	// Watch for changes to DaemonSet
+	return c.Watch(source.Kind(mgr.GetCache(), &appsv1.DaemonSet{}), &handler.EnqueueRequestForObject{}, predicate.Funcs{UpdateFunc: updateHandler})
```

Comment:
in the predicate, can we filter for daemonset that has certain annotations, e.g. DaemonSetPartitionAnnotation